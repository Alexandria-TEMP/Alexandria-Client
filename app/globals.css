@tailwind base;
@tailwind components;
@tailwind utilities;

@layer utilities {
  .text-balance {
    text-wrap: balance;
  }
}

@layer base {
  h1 {
    @apply text-2xl font-bold;
  }
  h2 {
    @apply text-xl font-semibold;
  }
  h3 {
    @apply text-lg;
  }
  h6 {
    @apply text-sm;
  }
}
