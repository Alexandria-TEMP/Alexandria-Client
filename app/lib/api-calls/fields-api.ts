import useSWR, { SWRResponse } from "swr";
import { ScientificFieldT } from "../types/api-types";
import { baseUrl, validateResponse } from "./api-common";

/**
 * Hook that fetches all scientific fields from the database, using SWR for caching, loading and error states
 * Fetching all tags from the database sounds like a very bad idea, but we would need some kind of lazy loading for MultiSelectAutocomplete otherwise
 * that is currently in todo
 * @returns the scientific field objects as array, as well as other SWR states (loading, error)
 */
export function useScientificFields(): SWRResponse<ScientificFieldT[], Error> {
  return useSWR(baseUrl + "/tags/scientific", async (...args) => {
    const response = await fetch(...args);
    await validateResponse(response);
    //disable reason: idk how to fix this and still get the correct type cause typescript
    // i have to look into this
    // eslint-disable-next-line @typescript-eslint/no-unsafe-assignment
    const scientificFieldTags: ScientificFieldT[] = await response.json();
    return scientificFieldTags;
  });
}
