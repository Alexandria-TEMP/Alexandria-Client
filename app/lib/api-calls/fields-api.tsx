// methods in this file will be heavily changed once integration with back end it done
// just retrieve some dummy data for now

export async function getFields() {
  await new Promise((resolve) => setTimeout(resolve, 1000));
  const data = [
    {
      id: "1",
      tag: "Computer Science",
      tagType: "ScientificField",
    },
    {
      id: "2",
      tag: "Mathematics",
      tagType: "ScientificField",
    },
    {
      id: "3",
      tag: "Psychology & Psychiatry",
      tagType: "ScientificField",
    },
    {
      id: "4",
      tag: "Medicine",
      tagType: "ScientificField",
    },
  ];

  return data;
}
