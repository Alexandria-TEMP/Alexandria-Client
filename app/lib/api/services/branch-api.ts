import { BranchUnionT, idBranchUnionT } from "@/lib/types/branch-union";
import { BranchT, ClosedBranchT, idT } from "../../types/api-types";
import { baseUrl, validateResponse } from "../api-common";
import fetchPostData from "./post-api";

/**
 * Fetches branch or closed branch data in a unified object
 * @async
 * @param id branch or closed branch ID
 * @returns branch and optionally closed branch data
 */
export async function fetchBranchData(
  id: idBranchUnionT,
): Promise<BranchUnionT> {
  let closedBranch = undefined;

  if (id.isClosed) {
    const closedBranchResponse = await fetch(
      `${baseUrl}/branches/closed/${id.id}`,
    );
    await validateResponse(closedBranchResponse);
    closedBranch = (await closedBranchResponse.json()) as ClosedBranchT;
  }

  const branchID = closedBranch?.branchID ?? (id.id as idT);
  const branchResponse = await fetch(`${baseUrl}/branches/${branchID}`);
  await validateResponse(branchResponse);

  const branch = (await branchResponse.json()) as BranchT;

  const projectPostID = branch.projectPostID ?? closedBranch?.projectPostID;
  if (!projectPostID) {
    throw new Error(
      `missing some project post ID in ${id.isClosed ? "closed" : ""} branch ${id.id}`,
    );
  }

  const updated = await fetchBranchUpdatedFieldsFallback(branch, projectPostID);

  return { branch, closedBranch, updated, projectPostID };
}

/**
 * For each of the possible post fields that a branch updates, returns
 * either the updated data or the current data if updated is null
 * @param branch branch whose update we're interested in
 * @param projectPostID ID of project post that branch is updating
 */
export async function fetchBranchUpdatedFieldsFallback(
  branch: BranchT,
  projectPostID: idT,
) {
  if (
    branch.updatedPostTitle &&
    branch.updatedCompletionStatus &&
    branch.updatedScientificFieldTagContainerID
  ) {
    return {
      postTitle: branch.updatedPostTitle,
      completionStatus: branch.updatedCompletionStatus,
      scientificFieldTagContainerID:
        branch.updatedScientificFieldTagContainerID,
    };
  }

  const postData = await fetchPostData({
    id: projectPostID,
    isProject: true,
  });

  return {
    postTitle: branch.updatedPostTitle ?? postData.post.title,
    completionStatus:
      branch.updatedCompletionStatus ??
      postData.projectPost!.projectCompletionStatus,
    scientificFieldTagContainerID:
      branch.updatedScientificFieldTagContainerID ??
      postData.post.scientificFieldTagContainerID,
  };
}

/**
 * Gets branches of a post given their ID.
 * @async
 * @param id Post ID
 */
// TODO remove next line
// eslint-disable-next-line @typescript-eslint/no-unused-vars
export async function getPostBranches(id: idT) {
  await new Promise((resolve) => setTimeout(resolve, 100));
  // TODO
  return {
    rejected: [1],
    open: [2],
    accepted: [3],
  };
}

/**
 * TODO jsdoc when properly implemented
 */
// TODO remove next line
// eslint-disable-next-line @typescript-eslint/no-unused-vars
export async function getBranchReviewStatuses(id: idT) {
  // TODO
  await new Promise((resolve) => setTimeout(resolve, 70));
  return [];
  // if (id == 1) return ["accept", "reject", "accept"];
  // else if (id == 2) return ["accept", "open", "open"];
  // else return ["accept", "accept", "accept"];
}
