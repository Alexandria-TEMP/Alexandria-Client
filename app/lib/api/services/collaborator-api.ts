/* eslint-disable @typescript-eslint/no-unused-vars */
// TODO remove previous line once implemented

import { idT } from "../../types/api-types";
import { QuartoContainerTypeT } from "../../types/quarto-container";
import { baseUrl } from "../api-common";

/**
 * Builds URL path for collaborator API calls
 * @param id collaborator ID
 * @param containerType if Quarto project is in a post or branch
 */
const buildResourcePath = (id: idT, containerType: QuartoContainerTypeT) =>
  `${baseUrl}/${containerType === "branch" ? "branches" : "posts"}/collaborators/${id}`;

/**
 * TODO
 */
export async function fetchCollaborators(
  ids: idT[],
  containerType: QuartoContainerTypeT,
) {
  // TODO
}
