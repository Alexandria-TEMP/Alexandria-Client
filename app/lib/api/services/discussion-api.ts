import { DiscussionContainerT, idT } from "@/lib/types/api-types";
import { baseUrl, validateResponse } from "../api-common";

/**
 * Get a discussion container by its ID, to access its discussions
 * @param id discussion container ID
 */
export async function fetchDiscussionContainer(
  id: idT,
): Promise<DiscussionContainerT> {
  const res = await fetch(`${baseUrl}/discussion-containers/${id}`);
  await validateResponse(res);
  return (await res.json()) as DiscussionContainerT;
}

/**
 * Gets data for a discussion given its ID.
 * @async
 * @param id Discussion ID
 */
export async function getDiscussionData(id: string) {
  // TODO
  // const res = await fetch();
  await new Promise((resolve) => setTimeout(resolve, 100));

  return {
    id: id,
    anonymous: false,
    author: {
      email: "mariecurie@tudelft.nl",
      firstName: "Marie",
      institution: "TU Delft",
      lastName: "Curie",
    },
    createdAt: "11 May 2024",
    deleted: false,
    deletedAt: "-",
    replies: [],
    text: "Lorem ipsum dolor sit amet, consectetur adipisicing elit. Sunt ut ducimus amet ex qui eius corrupti reiciendis, quibusdam suscipit, aspernatur ipsum. Reprehenderit libero molestias nostrum eum sed? Illo, quidem ad.",
    updatedAt: "11 May 2024",
  };
}

// TODO parameter type should include all data needed for a discussion
/**
 * Creates a new Discussion for a given Version.
 * @async
 * @param text Discussion contents
 * @param id discussion container ID
 */
export async function uploadDiscussion(text: string, id: idT) {
  // TODO
  // should include author, possible different discussion to reply to (and anything else that I'm forgetting)
  await new Promise((resolve) => setTimeout(resolve, 100));
  return `uploaded ${text} under id ${id}`;
}
