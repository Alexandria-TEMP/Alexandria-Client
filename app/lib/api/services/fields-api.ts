import {
  ScientificFieldTagContainerT,
  ScientificFieldTagT,
  idT,
} from "../../types/api-types";
import { baseUrl, validateResponse } from "../api-common";

/**
 * TODO jsdoc when properly implemented
 */
export async function getFields() {
  // TODO
  await new Promise((resolve) => setTimeout(resolve, 1000));
  const data = [
    {
      id: "1",
      tag: "Computer Science",
      tagType: "ScientificField",
    },
    {
      id: "2",
      tag: "Mathematics",
      tagType: "ScientificField",
    },
    {
      id: "3",
      tag: "Psychology & Psychiatry",
      tagType: "ScientificField",
    },
    {
      id: "4",
      tag: "Medicine",
      tagType: "ScientificField",
    },
  ];

  return data;
}

/**
 * Fetches all scientific fields based on their IDs
 * @param ids scientific field IDs
 */
export async function fetchScientificFields(
  ids: idT[],
): Promise<ScientificFieldTagT[]> {
  let tags: ScientificFieldTagT[] = [];

  for (const id of ids) {
    const res = await fetch(`${baseUrl}/tags/scientific/${id}`);
    if (!res.ok) continue;
    const scientificField = (await res.json()) as ScientificFieldTagT;
    tags = [...tags, scientificField];
  }

  return tags;
}

/**
 * Fetch scientific field tag container
 * @param id of the container
 * @returns container object
 */
export async function fetchScientificFieldContainer(
  id: idT,
): Promise<ScientificFieldTagContainerT> {
  const res = await fetch(`${baseUrl}/tags/scientific/containers/${id}`);
  await validateResponse(res);
  const container = (await res.json()) as ScientificFieldTagContainerT;
  return container;
}

/**
 * Method that gets the scientific fiels from a container
 * @param id container id
 * @returns a list of scientific field tags in that container
 */
export async function fetchScientificFieldsFromContainer(
  id: idT,
): Promise<ScientificFieldTagT[]> {
  const container = await fetchScientificFieldContainer(id);
  return await fetchScientificFields(container.scientificFieldTagIDs);
}
