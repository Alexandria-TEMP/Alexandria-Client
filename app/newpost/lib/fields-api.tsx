// methods in this file will be heavily changed once integration with back end it done
// just retrieve some dummy data for now

export function getFieldData(id: string) {
  // TODO
  if (id == "1") {
    return {
      id: "1",
      tag: "Computer Science",
      tagType: "ScientificField",
    };
  } else if (id == "2") {
    return {
      id: "2",
      tag: "Mathematics",
      tagType: "ScientificField",
    };
  } else if (id == "3") {
    return {
      id: "3",
      tag: "Psychology & Psychiatry",
      tagType: "ScientificField",
    };
  } else {
    return {
      id: "4",
      tag: "Medicine",
      tagType: "ScientificField",
    };
  }
}

export function getFields() {
  // TODO
  const data = [
    {
      id: "1",
      tag: "Computer Science",
      tagType: "ScientificField",
    },
    {
      id: "2",
      tag: "Mathematics",
      tagType: "ScientificField",
    },
    {
      id: "3",
      tag: "Psychology & Psychiatry",
      tagType: "ScientificField",
    },
    {
      id: "4",
      tag: "Medicine",
      tagType: "ScientificField",
    },
  ];

  return new Map(data.map((m) => [m.id, m]));
}
