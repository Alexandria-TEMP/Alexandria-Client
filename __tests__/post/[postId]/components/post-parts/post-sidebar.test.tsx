import fetchPostData from "@/lib/api/services/post-api";
import PostSidebar from "@/post/[postId]/components/post-parts/post-sidebar";
import { expect, describe, it } from "@jest/globals";
import { render } from "@testing-library/react";
import {
  dummyPostUnion,
  dummyScientificField,
} from "~/__tests__/__utils__/dummys";
import AuthorCardList from "@/post/[postId]/components/cards/author-card-list";
import { fetchScientificFieldsFromContainer } from "@/lib/api/services/fields-api";
import { fetchPostCollaboratorsAsSortedMemberIDs } from "@/lib/api/services/collaborator-api";

jest.mock("@/lib/api/services/post-api");
jest.mock("@/lib/api/services/fields-api");
jest.mock("@/lib/api/services/collaborator-api");
jest.mock("@/post/[postId]/components/cards/author-card-list");

describe("PostSidebar", () => {
  (fetchPostData as jest.Mock).mockResolvedValue(dummyPostUnion.withProject);
  (fetchPostCollaboratorsAsSortedMemberIDs as jest.Mock).mockResolvedValue({
    author: [1],
    contributor: [1],
    reviewer: [1],
  });
  (fetchScientificFieldsFromContainer as jest.Mock).mockResolvedValue([
    dummyScientificField,
    dummyScientificField,
    dummyScientificField,
  ]);
  (AuthorCardList as jest.Mock).mockReturnValue(<p>List of members</p>);

  it("matches snapshot", async () => {
    const { container } = render(await PostSidebar({ id: 1, isProject: true }));
    expect(container).toMatchSnapshot();
  });
});
