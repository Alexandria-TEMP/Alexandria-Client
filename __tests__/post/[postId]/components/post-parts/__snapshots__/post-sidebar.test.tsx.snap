// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PostSidebar matches snapshot 1`] = `
<div>
  <div
    class="sticky top-20"
  >
    <h3>
      Scientific fields
    </h3>
    <div
      class="flex flex-row flex-wrap gap-3"
    >
      <div
        class="relative max-w-fit min-w-min inline-flex items-center justify-between box-border whitespace-nowrap px-1 h-7 text-small rounded-full bg-default text-default-foreground"
      >
        <span
          class="flex-1 text-inherit font-normal px-2"
        >
          Computer Science
        </span>
      </div>
      <div
        class="relative max-w-fit min-w-min inline-flex items-center justify-between box-border whitespace-nowrap px-1 h-7 text-small rounded-full bg-default text-default-foreground"
      >
        <span
          class="flex-1 text-inherit font-normal px-2"
        >
          Mathematics
        </span>
      </div>
      <div
        class="relative max-w-fit min-w-min inline-flex items-center justify-between box-border whitespace-nowrap px-1 h-7 text-small rounded-full bg-default text-default-foreground"
      >
        <span
          class="flex-1 text-inherit font-normal px-2"
        >
          Theory of computation
        </span>
      </div>
    </div>
    <div
      class="h-4"
    />
    <h3>
      Authors
    </h3>
    <p>
      List of members
    </p>
    <div
      class="h-4"
    />
    <h3>
      Collaborators
    </h3>
    <p>
      List of members
    </p>
    <div
      class="h-4"
    />
  </div>
</div>
`;
